docs/usage/model_config.md

```diff
@@ -164,3 +167,31 @@ For example:
 {!.tmp_examples/model_config_alias_precedence.py!}
 ```
 _(This script is complete, it should run "as is")_
+
+## Smart Union
+
+By default, as explained [here](types.md#unions), _pydantic_ tries to validate (and coerce if it can) in the order of the `Union`.
+So sometimes you may have unexpected coerced data.
+
+```py
+{!.tmp_examples/model_config_smart_union_off.py!}
+```
+_(This script is complete, it should run "as is")_
+
+To prevent this, you can enable `Config.smart_union`. _Pydantic_ will then check all allowed types before even trying to coerce.
+Know that this is of course slower, especially if your `Union` is quite big.
+
+```py
+{!.tmp_examples/model_config_smart_union_on.py!}
+```
+_(This script is complete, it should run "as is")_
+
+!!! warning
+
+Note that this option **does not support compound types yet** (e.g. differentiate `List[int]` and `List[str]`).
```

Comment:
I think this is supposed to be the warning text, so should be indented, see [here](https://smokeshow.helpmanual.io/373p4j603c2v6f615z0y/usage/model_config/#smart-union):

![image](https://user-images.githubusercontent.com/4039449/144746982-b2ac9ae2-ebc4-4771-89f3-3bfa0cd2e653.png)